zson/formatter.go

```diff
@@ -0,0 +1,401 @@
+package zson
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zng"
+)
+
+type Formatter struct {
+	typedefs    typemap
+	tab         int
+	newline     string
+	whitespace  string
+	typeTab     int
+	typeNewline string
+	nid         int
+	types       *TypeTable
+	builder     strings.Builder
+	stack       []strings.Builder
+}
+
+func NewFormatter(pretty int) *Formatter {
+	newline := ""
+	if pretty > 0 {
+		newline = "\n"
+	}
+	return &Formatter{
+		typedefs:   make(typemap),
+		tab:        pretty,
+		newline:    newline,
+		whitespace: strings.Repeat(" ", 80),
+	}
+}
+
+func (f *Formatter) push() {
+	f.stack = append(f.stack, f.builder)
+	f.builder = strings.Builder{}
```

Comment:
```suggestion
	f.builder.Reset()
```

Reply:
I actually first wrote the code this way then wanted to make sure it works, and while it does work, the docs for strings.Builder say to not make a copy without further elaboration.  I take this to mean that you definitely shouldn't write into the copy but Reset happens to work because it just zeros out the pointers, which is the same as what I'm doing here but this way conforms with the documentation.